Add blog post on improving cloud efficiency (#723)
ServiceWeaver/weaver: website/blog/improve_cloud_efficiency.md

# Improve Cloud Efficiency with Service Weaver

<div class="blog-author">Shane Ouchi</div>
<div class="blog-date">February 7th, 2024</div>

## Overview

<div style='
    background-color: #FFF6CF;
    border-radius: 16px;
    padding:0.7em;
    margin-top: 1rem;
    border-width: 1px;
    border-color: black;
    border-style: none;'>
<p style='
    font-size: 1.2em;
    text-align: left;
    margin-top: 0.5em;
    color: #e06666';><b><i>
📝 Service Weaver provides a simplified, safe and scalable way to improve
backend cloud application efficiency for organizations looking to reduce their
cloud spend.</i></b></p>
</div>

Currently, organizations top challenge is [managing cloud spending][1] with IT leaders
expecting [31% of their operating costs][2] going to the cloud and over [⅔ expecting
YoY increases in cloud spending][2]. One way that organizations can reduce their cloud
spending is by improving the efficiency of developing and maintaining their cloud

> applications. Organizations can evaluate 3 key areas of inefficiencies: 1/ Development

Reviewer: ```suggestion
applications. Organizations can evaluate three key areas of inefficiencies: 1/ Development
```

Author: Done

> applications. Organizations can evaluate 3 key areas of inefficiencies: 1/ Development

Reviewer: Mind replacing the 1/, 2/, 3/ bullet points with markdown bullet points like:

1. Development...
2. IT burden...
3. Compute...

Author: Done

> overhead that negatively impacts feature velocity; 2/  IT burden to setup & maintain

Reviewer: ```suggestion
overhead that negatively impacts feature velocity; 2/  IT burden to setup and maintain
```

Author: Done

apps for secure, reliable environments; 3/ Compute waste that over-allocates compute
resources needed to run globally available applications.  Service Weaver provides
an effective development solution that can holistically improve these efficiency
challenges across the board.

## Simplify development overhead

> Service Weaver can simplify development overhead with a flexible, easy-to-use
> framework that enables development teams to focus on providing differentiating value.

Organizations continue to focus on making teams more effective at balancing
engineering productivity, product quality and differentiating features. Microservices
have made this more difficult by adding overhead for development teams, who increasingly
spend their time on planning, testing, tech debt and incident response.

> According to the most recent [Go Developer survey][3], “testing and debugging [is]
> the biggest challenge when writing microservice-based applications, followed by
> operational complexity.” Teams also report high upfront cognitive load on

Reviewer: ```suggestion
According to the most recent [Go Developer survey][3], "testing and debugging [is]
the biggest challenge when writing microservice-based applications, followed by
operational complexity." Teams also report high upfront cognitive load on
```

containerized runtimes raise the abstraction layer of computing efficiency to the
pod/container level. They provide operator teams with configuration knobs to
auto-scale based on average pod-level utilization metrics and balance traffic
intra- or inter-region using default pod metrics. Depending on the traffic
distribution through an application, this can still leave a "fat-tail” of pods
that are highly underutilized.

To minimize computing waste, teams need to holistically evaluate: 1/ lower latency
networking, 2/ custom load balancing metrics, 3/ improved sharding capabilities,
and 4/ enhanced aggregation mechanisms for horizontal auto-scaling that optimizes
app-specific configuration. More importantly, they want to interact with these
capabilities from a highly simplified interface.

Service Weaver provides an easy-to-use framework with built-in primitives to
address scaling inefficiencies with up to 9x reduction in compute waste
out-of-the-box (*Figure 5*). Service Weaver also provides advanced capabilities
e.g. inter-component communication (*Figure 4*) to further evaluate and fine-tune
efficiencies according to your scalability or performance needs to enable operators

> to experiment & minimize compute waste with minimal effort.

Reviewer: ```suggestion
to experiment and minimize compute waste with minimal effort.
```

We plan to go further in-depth on this topic in an upcoming post.

![integrated](../assets/images/faster_grpc.png)
**Figure 4**: *Service Weaver has an advanced inter-component communication
protocol that is up to **3X** faster than [gRPC][grpc].*

<br>

<p align="center">
<img src="../assets/images/benchmark.png" alt="benchmark" width="60%">
</p>

**Figure 5**: *Benchmarking an 11-microservice [Online Boutique][16] app with a
rewritten 11-component Service Weaver modulith on [GKE][gke] resulted in an overall **9X**
reduction in CPUs when loading ~10K queries per second.  See our [paper][17] for mode
details on our benchmarking.*

## How to get started?

* To get started on your first app, we recommend following this [step-by-step][6] guide.

> * To evaluate your application in a staging or production environment, you can
> * take the [multiprocess execution][7] and [deploy it to Cloud Run][8], [GKE][9] or [Kubernetes][10].

Reviewer: ```suggestion
* To evaluate your application in a staging or production environment, you can take the [multiprocess executor][7] and deploy your app to [Cloud Run][8], [GKE][9] or [Kubernetes][10].
```

# Improve Cloud Efficiency with Service Weaver

<div class="blog-author">Shane Ouchi</div>
<div class="blog-date">February 8th, 2024</div>

## Overview

<div style='
    background-color: #FFF6CF;
    border-radius: 16px;
    padding:0.7em;
    margin-top: 1rem;
    border-width: 1px;
    border-color: black;
    border-style: none;'>
<p style='
    font-size: 1.2em;
    text-align: left;
    margin-top: 0.5em;
    color: #e06666'><b><i>
📝 Service Weaver provides a simplified, safe and scalable way to improve
backend cloud application efficiency for organizations looking to reduce their
cloud spend.</i></b></p>
</div>

Currently, organizations top challenge is [managing cloud spending][1] with IT leaders
expecting [31% of their operating costs][2] going to the cloud and over [⅔ expecting
YoY increases in cloud spending][2]. One way that organizations can reduce their cloud
spending is by improving the efficiency of developing and maintaining their cloud
applications. Organizations can evaluate three key areas of inefficiencies:

1. Development overhead that negatively impacts feature velocity;
2. IT burden to setup and maintain apps for secure, reliable environments;

> 3. Compute waste that over-allocates compute resources needed to run globally
> available applications.  Service Weaver provides an effective development solution
> that can holistically improve these efficiency challenges across the board.

Reviewer: ```suggestion
3. Compute waste that over-allocates compute resources needed to run globally
available applications.  

Service Weaver provides an effective development solution
that can holistically improve these efficiency challenges across the board.
```

Author: Thanks

Organizations continue to deepen adoption of public cloud platforms (e.g. [GCP][gcp])
to offload IT burden when running reliable applications. Teams need to maintain
feature velocity while meeting organizational security and compliance needs.

One of the [top pain points][4] preventing cloud adoption is simplifying authentication
and security needs while maintaining compliance with regulatory requirements. End
users want confidence that their data is protected in-transit and explicitly used
for intended purposes; while operator teams need simpler ways to remediate incidents
and prevent misuse. Platform / IT teams continue growing the list of networking
tools and policies to make connecting and routing traffic between services more
reliable and secure. This has resulted in a higher IT burden across platform,
operator and development teams.

Service Weaver can minimize IT burden for teams who are looking for an easy-to-use
way to bootstrap a safer app on public cloud platforms. Once apps are bootstrapped,
applications will have default capabilities to:

1. Ensure data in-transit is secure with [mTLS][mtls];
2. Reduce cascading failures with fault isolated components;

> 3. Automate [IAM][iam], logging and monitoring when deploying to [Google Cloud][gcloud] (*Figure 3*).
> Teams can also choose to export data to open-source observability providers such
> as [Prometheus][prometheus] and [Grafana][grafana].

Reviewer: ```suggestion
3. Automate [IAM][iam], logging and monitoring when deploying to [Google Cloud][gcloud] (*Figure 3*).

Teams can also choose to export data to open-source observability providers such
as [Prometheus][prometheus] and [Grafana][grafana].
```

out-of-the-box (*Figure 5*). Service Weaver also provides advanced capabilities
e.g. inter-component communication (*Figure 4*) to further evaluate and fine-tune
efficiencies according to your scalability or performance needs to enable operators
to experiment and minimize compute waste with minimal effort.

We plan to go further in-depth on this topic in an upcoming post.

![integrated](../assets/images/faster_grpc.png)
**Figure 4**: *Service Weaver has an advanced inter-component communication
protocol that is up to **3X** faster than [gRPC][grpc].*

<br>

<p align="center">
<img src="../assets/images/benchmark.png" alt="benchmark" width="60%">
</p>

**Figure 5**: *Benchmarking an 11-microservice [Online Boutique][16] app with a
rewritten 11-component Service Weaver modulith on [GKE][gke] resulted in an overall **9X**
reduction in CPUs when loading ~10K queries per second.  See our [paper][17] for mode
details on our benchmarking.*

## How to get started?

* To get started on your first app, we recommend following this [step-by-step][6] guide.
* To evaluate your application in a staging or production environment, you can

> take the [multiprocess execution][7] and [deploy it to Cloud Run][8], [GKE][9] or [Kubernetes][10].

Reviewer: ```suggestion
take the [multiprocess executor][7] and [deploy it to Cloud Run][8], [GKE][9] or [Kubernetes][10].
```
